# SQL

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Задание-№1" data-toc-modified-id="Задание-№1-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Задание №1</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Задание-№2" data-toc-modified-id="Задание-№2-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Задание №2</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Задание-№3" data-toc-modified-id="Задание-№3-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задание №3</a></span><ul class="toc-item"><li><span><a href="#ВЫВОД:" data-toc-modified-id="ВЫВОД:-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>ВЫВОД:</a></span></li></ul></li><li><span><a href="#Задание-№4" data-toc-modified-id="Задание-№4-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Задание №4</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Задание-№5" data-toc-modified-id="Задание-№5-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Задание №5</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Цель исследования   

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Цель этого исследования - использовать набор данных, предоставленный одним из сервисов, конкурирующих на рынке книжных приложений, для создания ценностного предложения для нового продукта.**

# Задание  
- Посчитайте, сколько книг вышло после 1 января 2000 года;  
- Для каждой книги посчитайте количество обзоров и среднюю оценку;  
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;  
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;  
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

# Описание данных  

**Таблица books**

Содержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

**Таблица authors**

Содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.

**Таблица publishers**

Содержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;

**Таблица ratings**

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

**Таблица reviews**

Содержит данные о пользовательских обзорах на книги:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
- text — текст обзора.

# Схема данных

![](https://pictures.s3.yandex.net/resources/scheme_1589269096.png)

# Структура базы данных:

In [1]:
from IPython.display import Image

Исследуйте таблицы

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
import datetime as dt

In [3]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db']) 

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

for table in ['books', 'authors', 'publishers', 'ratings', 'reviews']:
    query = ''' SELECT * FROM {}; 
    '''.format(table) 

    request = pd.io.sql.read_sql(query, con = engine) 
    display(request.head())
    print(request.info())
    print(f'Таблица - "{table}"')

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
None
Таблица - "books"


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
None
Таблица - "authors"


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
None
Таблица - "publishers"


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
None
Таблица - "ratings"


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
None
Таблица - "reviews"


- Данные в порядке, пропусков нет

Запишем каждую таблицу в ДатаФрейм

In [4]:
query_books = ''' SELECT * FROM books; ''' 
books = pd.io.sql.read_sql(query_books, con = engine)
# books

query_reviews = ''' SELECT * FROM reviews; ''' 
reviews = pd.io.sql.read_sql(query_reviews, con = engine)
# reviews

query_ratings = ''' SELECT * FROM ratings; ''' 
ratings = pd.io.sql.read_sql(query_ratings, con = engine)
# ratings

query_publishers = ''' SELECT * FROM publishers; ''' 
publishers = pd.io.sql.read_sql(query_publishers, con = engine)
# publishers

query_authors = ''' SELECT * FROM authors; ''' 
authors = pd.io.sql.read_sql(query_authors, con = engine)
# authors

## Задание №1  

Посчитаем, сколько книг вышло после 1 января 2000 года

In [5]:
query1 = ''' 
SELECT COUNT(*) 
FROM books
WHERE publication_date > '2000-1-1'; 
''' 

request1 = pd.io.sql.read_sql(query1, con = engine)
# request1

print(f'После 1 января 2000 года вышло {request1.iloc[0,0]} книг')

После 1 января 2000 года вышло 819 книг


In [6]:
# check
books['publication_date'] = books['publication_date'].astype('datetime64')
books[books['publication_date']>'2000-01-01']['book_id'].count()

819

### Вывод:  
- **После 1 января 2000 года вышло 819 книг**

## Задание №2  

Для каждой книги посчитаем количество обзоров и среднюю оценку

In [7]:
query2 = ''' 

SELECT 
    books.title AS book_name,
    books.book_id AS id,
    SubRev.n_rev AS reviews,
    SubRat.avg_rat AS avg_rating
FROM books LEFT JOIN 
                (SELECT
                    book_id,
                    AVG(rating) AS avg_rat
                FROM
                    ratings
                GROUP BY 
                    book_id
                ) AS SubRat ON SubRat.book_id = books.book_id
        LEFT JOIN
                (SELECT 
                    book_id,
                    COUNT(review_id) AS n_rev
                FROM
                    reviews
                GROUP BY 
                    book_id
                ) AS SubRev ON SubRev.book_id = books.book_id
                
; ''' 

request2 = pd.io.sql.read_sql(query2, con = engine)
print('Количество обзоров и средняя оценка:')
request2

Количество обзоров и средняя оценка:


book_name   id  reviews  \
0            The Body in the Library (Miss Marple  #3)  652      2.0   
1                                            Galápagos  273      2.0   
2                             A Tree Grows in Brooklyn   51      5.0   
3    Undaunted Courage: The Pioneering First Missio...  951      2.0   
4                                          The Prophet  839      4.0   
..                                                 ...  ...      ...   
995                                Alice in Wonderland   64      4.0   
996          A Woman of Substance (Emma Harte Saga #1)   55      2.0   
997                                          Christine  148      3.0   
998  The Magicians' Guild (Black Magician Trilogy  #1)  790      2.0   
999                           The Plot Against America  828      2.0   

     avg_rating  
0      4.500000  
1      4.500000  
2      4.250000  
3      4.000000  
4      4.285714  
..          ...  
995    4.230769  
996    5.000000  
997    3.428571  
998    3.500000  
999    3.000000  

[1000 rows x 4 columns]

In [8]:
# check
r2 = books[['title','book_id']].merge(reviews.groupby('book_id')['review_id'].count().reset_index(), 
                                 how='left', on='book_id').merge(ratings.groupby('book_id')['rating'].mean()\
                                                                 .reset_index(), how='left', on='book_id')
r2.query('book_id == [652,273,51,951,839,64,55,148,790,828]')

title  book_id  review_id  \
50                            A Tree Grows in Brooklyn       51        5.0   
54           A Woman of Substance (Emma Harte Saga #1)       55        2.0   
63                                 Alice in Wonderland       64        4.0   
147                                          Christine      148        3.0   
272                                          Galápagos      273        2.0   
651          The Body in the Library (Miss Marple  #3)      652        2.0   
789  The Magicians' Guild (Black Magician Trilogy  #1)      790        2.0   
827                           The Plot Against America      828        2.0   
838                                        The Prophet      839        4.0   
950  Undaunted Courage: The Pioneering First Missio...      951        2.0   

       rating  
50   4.250000  
54   5.000000  
63   4.230769  
147  3.428571  
272  4.500000  
651  4.500000  
789  3.500000  
827  3.000000  
838  4.285714  
950  4.000000

In [9]:
mean = request2['avg_rating'].mean()
print('Средняя оценка:', mean)

Средняя оценка: 3.8989730839448864


### Вывод:   
**Количество обзоров: от 51 до 951**  
**Средняя оценка: от 3.9**

## Задание №3   

Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — таким образом исключим из анализа брошюры

In [10]:
query3 = ''' 
SELECT
    T.publisher_id,
    T.publisher,
    T.n_books
FROM
    (SELECT 
        books.publisher_id AS publisher_id,
        COUNT(books.book_id) AS n_books,
        publishers.publisher AS publisher 
    FROM 
        books 
        LEFT JOIN publishers ON books.publisher_id=publishers.publisher_id

    WHERE 
        num_pages>50
    GROUP BY
        books.publisher_id,
        publishers.publisher
    ORDER BY 
        n_books DESC) AS T
LIMIT 1
; ''' 

request3 = pd.io.sql.read_sql(query3, con = engine)
print('Издательство, которое выпустило наибольшее число книг толще 50 страниц:')
request3

Издательство, которое выпустило наибольшее число книг толще 50 страниц:


publisher_id      publisher  n_books
0           212  Penguin Books       42

In [11]:
# check
books.query('num_pages>50')[['publisher_id','num_pages', 'book_id']]. \
        groupby('publisher_id')['book_id'].count().reset_index().sort_values(by='book_id',ascending=False)

publisher_id  book_id
207           212       42
302           309       31
111           116       25
212           217       24
31             33       19
..            ...      ...
83             86        1
199           204        1
200           205        1
82             85        1
333           340        1

[334 rows x 2 columns]

### ВЫВОД:   
**Издательство, которое выпустило наибольшее число книг, 42 книги, толще 50 страниц - это "Penguin Books"**

## Задание №4  

Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками

In [12]:
query4 = ''' 
SELECT 
    aut_avg.author,
    AVG(aut_avg.avg_rating) AS avg_avg_rating
FROM
        (SELECT 
            authors.author AS author
           , AVG(ratings.rating) AS avg_rating
        FROM books
            LEFT JOIN authors ON books.author_id = authors.author_id
            LEFT JOIN ratings ON books.book_id = ratings.book_id
        GROUP BY 
             books.book_id
             , authors.author_id
        HAVING
            COUNT(ratings.rating_id) >=50 
        ) AS aut_avg 
GROUP BY    
    aut_avg.author
ORDER BY
    avg_avg_rating DESC
LIMIT 5
    
; ''' 

request4 = pd.io.sql.read_sql(query4, con = engine)
print('Авторы с самой высокой средней оценкой книг:')
request4

Авторы с самой высокой средней оценкой книг:


author  avg_avg_rating
0         J.K. Rowling/Mary GrandPré        4.283844
1  Markus Zusak/Cao Xuân Việt Khương        4.264151
2                     J.R.R. Tolkien        4.258446
3                  Louisa May Alcott        4.192308
4                       Rick Riordan        4.080645

### Вывод:   
**Самую высокую среднюю оценку получает - J.K. Rowling/Mary GrandPré - 4.283844 (Гарри Поттер ФОРЕВЕР)**

## Задание №5   

Посчитаим cреднее количество текстовых отзывов среди пользователей, оценивших более 50 книг  

Из предоставленных данных непонятно являются ли пользователи из таблиц ratings и reviews одними и теми же, провериим

In [13]:
print('Кол-во пользователей в "reviews" =', reviews['username'].nunique())
print('Кол-во пользователей в "ratings" =', ratings['username'].nunique())

Кол-во пользователей в "reviews" = 160
Кол-во пользователей в "ratings" = 160


По кол-ву совпадает, но проверим еще по множествам

In [14]:
set(reviews['username'].unique().tolist()) == set(ratings['username'].unique().tolist())

True

Множества идентичны - это одни и теже пользователи  
Составляем запрос

In [15]:
query5 = ''' 

SELECT
    AVG(Sub.n_reviews)
FROM
    (SELECT
        username,
        COUNT(review_id) AS n_reviews
    FROM
        reviews
    WHERE 
        username IN (
                SELECT 
                    username
                FROM 
                    ratings
                GROUP BY 
                    username
                HAVING 
                    COUNT(rating_id) > 50
                    )
    GROUP BY 
        username
    ) AS Sub
; ''' 

request5 = pd.io.sql.read_sql(query5, con = engine)
request5

avg
0  24.333333

In [16]:
# check
users = ratings.groupby('username')['rating_id'].count().reset_index()
users = users.query('rating_id > 50')
users

username  rating_id
57   jennifermiller         53
83       martinadam         56
104          paul88         56
110       richard89         55
123     sfitzgerald         55
153          xdavis         51

In [17]:
users_lst = users["username"].tolist()
print('Среднее количество текстовых отзывов среди пользователей, оценивших более 50 книг, составляет:',\
      reviews.query('username in @users_lst').groupby('username')['review_id'].count().mean())

Среднее количество текстовых отзывов среди пользователей, оценивших более 50 книг, составляет: 24.333333333333332


### Вывод:  
**Среднее количество текстовых отзывов среди пользователей, оценивших более 50 книг, составляет: 24.33**

## Вывод   
- Количество книг, вышедших после 1 января 2000 года - **819**;  
- Количество обзоров: от 51 до 951, средняя оценка: **3.9**;  
- Издательство, которое выпустило наибольшее число книг толще 50 страниц — **Penguin Books**;  
- Автор с самой высокой средней оценкой книг при условии 50 и более оценок:   
  **J.K. Rowling/Mary GrandPré - 4.28 (Гарри Поттер ФОРЕВЕР)**  
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - **24.33**.